sw/airborne/modules/gps/gps_datalink.c

```diff
@@ -148,166 +113,109 @@ static void parse_gps_datalink_small(int16_t heading, uint32_t pos_xyz, uint32_t
   gps_datalink.last_3dfix_ticks = sys_time.nb_sec_rem;
   gps_datalink.last_3dfix_time = sys_time.nb_sec;
 
-  // publish new GPS data
   uint32_t now_ts = get_sys_time_usec();
+
+  // Publish GPS data
   AbiSendMsgGPS(GPS_DATALINK_ID, now_ts, &gps_datalink);
 }
 
-/** Parse the REMOTE_GPS datalink packet */
-static void parse_gps_datalink(uint8_t numsv, int32_t ecef_x, int32_t ecef_y, int32_t ecef_z,
-                               int32_t lat, int32_t lon, int32_t alt, int32_t hmsl,
-                               int32_t ecef_xd, int32_t ecef_yd, int32_t ecef_zd,
-                               uint32_t tow, int32_t course)
+/** Parse the full EXTERNAL_POSE message and publish as GPS through ABI */
+void gps_datalink_parse_EXTERNAL_POSE(uint8_t *buf)
 {
-  gps_datalink.lla_pos.lat = lat;
-  gps_datalink.lla_pos.lon = lon;
-  gps_datalink.lla_pos.alt = alt;
-  SetBit(gps_datalink.valid_fields, GPS_VALID_POS_LLA_BIT);
-
-  gps_datalink.hmsl        = hmsl;
-  SetBit(gps_datalink.valid_fields, GPS_VALID_HMSL_BIT);
+  if (DL_EXTERNAL_POSE_ac_id(buf) != AC_ID) { return; } // not for this aircraft
 
-  gps_datalink.ecef_pos.x = ecef_x;
-  gps_datalink.ecef_pos.y = ecef_y;
-  gps_datalink.ecef_pos.z = ecef_z;
-  SetBit(gps_datalink.valid_fields, GPS_VALID_POS_ECEF_BIT);
+  uint32_t tow = DL_EXTERNAL_POSE_timestamp(buf);
 
-  gps_datalink.ecef_vel.x = ecef_xd;
-  gps_datalink.ecef_vel.y = ecef_yd;
-  gps_datalink.ecef_vel.z = ecef_zd;
-  SetBit(gps_datalink.valid_fields, GPS_VALID_VEL_ECEF_BIT);
+  struct EnuCoor_f enu_pos, enu_speed;
+  enu_pos.x = DL_EXTERNAL_POSE_enu_x(buf);
+  enu_pos.y = DL_EXTERNAL_POSE_enu_y(buf);
+  enu_pos.z = DL_EXTERNAL_POSE_enu_z(buf);
+  enu_speed.x = DL_EXTERNAL_POSE_enu_xd(buf);
+  enu_speed.y = DL_EXTERNAL_POSE_enu_yd(buf);
+  enu_speed.z = DL_EXTERNAL_POSE_enu_zd(buf);
 
-  ned_of_ecef_vect_i(&gps_datalink.ned_vel, &ltp_def , &gps_datalink.ecef_vel);
-  SetBit(gps_datalink.valid_fields, GPS_VALID_VEL_NED_BIT);
+  struct FloatQuat body_q;
+  body_q.qi = DL_EXTERNAL_POSE_body_qi(buf);
+  body_q.qx = DL_EXTERNAL_POSE_body_qx(buf);
+  body_q.qy = DL_EXTERNAL_POSE_body_qy(buf);
+  body_q.qz = DL_EXTERNAL_POSE_body_qz(buf);
 
-  gps_datalink.gspeed = (int16_t)FLOAT_VECT2_NORM(gps_datalink.ned_vel);
-  gps_datalink.speed_3d = (int16_t)FLOAT_VECT3_NORM(gps_datalink.ned_vel);
+  struct FloatEulers body_e;
+  float_eulers_of_quat(&body_e, &body_q);
+  float heading = body_e.psi;
```

Comment:
The quaternion sent from optitrack is based on ENU frame, so the correct heading should be `-body_e.psi`. But if the source is sending a NED quat, this is correct. We have to decide if the quaternion is ENU or NED (hence converted onboard or on the ground).

Reply:
I now changed the `body_q` to NED. For me it was easiest now to just keep the transmitting in ENU. But if anyone is willing to throw away the old `natnet2ivy` and change it to NED I'm ok with that. I think it would be better if the ground sends it it in NED since almost everything onboard is done in NED.